File: 05/05.py
import pprint
from pathlib import Path

input = Path(__file__).parents[0] / "05.txt"

def stacks_moves():
    lines = input.read_text().splitlines()
    lines = iter(lines)

    stacks = [[] for _ in range(9)]

    for line in lines:
        if not line:
            break

        for i, element in enumerate(line[1::4]):
            if element != " ":
                stacks[i].append(element)

    for stack in stacks:
        stack.pop()
        stack.reverse()

    pprint.pprint(stacks)

    stacks = [None] + stacks

    def parse_moves():
        for move in lines:
            _, count, _, src, _, dst = move.split()
            count = int(count)
            src = int(src)
            dst = int(dst)
            yield (count, src, dst)

    moves = list(parse_moves())

    return stacks, moves

# part 1

stacks, moves = stacks_moves()

for count, src, dst in moves:
    for _ in range(count):
        stacks[dst].append(stacks[src].pop())

print("".join(stack[-1] for stack in stacks[1:]))

# part 2

stacks, moves = stacks_moves()

for count, src, dst in moves:
    leave = stacks[src][-count:]
    remain = stacks[src][:-count]
    assert len(leave) + len(remain) == len(stacks[src])
    stacks[src] = remain
    stacks[dst].extend(leave)
    # pprint.pprint(stacks)

print("".join(stack[-1] for stack in stacks[1:]))

# part 2 golf

stacks, moves = stacks_moves()

for count, src, dst in moves:
    floor = []
    for _ in range(count):
        floor.append(stacks[src].pop())
    while floor:
        stacks[dst].append(floor.pop())

print("".join(stack[-1] for stack in stacks[1:]))

File: 05/05.rs
#![feature(iter_array_chunks)]
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::vec::Vec;

fn main() {
    let mut lines = read_lines("./05/05.txt").unwrap();
    let mut stacks: Vec<String> = vec![];

    for line in &mut lines {
        let line = line.unwrap();
        if line.len() == 0 {
            break;
        }
        line[1..]
            .chars()
            .step_by(4)
            .enumerate()
            .for_each(|(i, chunk)| {
                if stacks.len() < i + 1 {
                    stacks.push(String::new());
                }
                // println!("{} {}", &i, &chunk);
                if chunk != ' ' {
                    stacks[i].insert(0, chunk);
                }
            });
    }

    for line in &mut lines {
        let unwrapped = line.unwrap();
        let moves: Vec<&str> = unwrapped.split(" ").collect();
        let count = moves[1].parse::<usize>().unwrap();
        let from = moves[3].parse::<usize>().unwrap();
        let to = moves[5].parse::<usize>().unwrap();
        // println!("{} {} {}", count, from, to);

        for _ in 0..count {
            let item = stacks[from - 1].pop().unwrap();
            stacks[to - 1].push(item);
        }
    }

    for i in 0..stacks.len() {
        print!("{}", stacks[i].pop().unwrap());
    }

    println!();
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}
